### PR TITLE
Fluentd 1.16 base image

### PR DESCRIPTION
**Pull request by OverOrion**
This PR introduces a three stages docker build:
- `base` => only the label router is added
- `minimal` => some plugins are added
- `full` => all plugins are added

I am more than happy to change which plugin goes where, just let me know.

Proof for the base image successfully running the E2E tests: https://github.com/OverOrion/logging-operator/pull/3

- [ ] The old v1.15 image is no longer in the CI, but the dockerfile is still there for it. Should I add back the old image, or just delete the whole folder?

### CODE REVIEW

File: v1.16/Dockerfile [hunk 1]
```diff
@@ -13,42 +19,74 @@ RUN apk update \
...
-    " \
  && apk add --no-cache --virtual .build-deps \
         build-base \
         ruby-dev gnupg \
- && apk add $buildDeps \
+ && apk add $BUILD_DEPS \
  && echo 'gem: --no-document' >> /etc/gemrc \
- && gem install bundler && bundle install \
- && apk del .build-deps $buildDeps \
+ && gem install fluentd -v ${FLUENTD_VERSION} \
```
Comment (pepov) [inline v1.16/Dockerfile:27]: isn't this already installed by the base image?

File: v1.16/Dockerfile [hunk 2]
```diff
@@ -13,42 +19,74 @@ RUN apk update \
...
+ARG BUILD_DEPS
+USER root
+# Do not split this into multiple RUN!
+# Docker creates a layer for every RUN-Statement
+# therefore an 'apk delete' has no effect
+RUN apk update \
+ && apk add --no-cache --virtual .build-deps \
+        build-base \
+        ruby-dev gnupg \
+ && apk add $BUILD_DEPS \
```
Comment (pepov) [inline v1.16/Dockerfile:61]: isn't this already installed in the base?

File: v1.16/Dockerfile [hunk 3]
```diff
@@ -13,42 +19,74 @@ RUN apk update \
...
+ && echo 'gem: --no-document' >> /etc/gemrc \
+ && gem install specific_install -v 0.3.8 \
```
Comment (pepov) [inline v1.16/Dockerfile:83]: isn't this already installed in minimal?

File: v1.16/Dockerfile [hunk 4]
```diff
@@ -13,42 +19,74 @@ RUN apk update \
...
-ENV RUBYLIB="/usr/lib/ruby/gems/3.1.0/gems/resolv-0.2.2/lib"
 
 ENV LD_PRELOAD=""
 EXPOSE 24224 5140
+USER fluent
+ 
+### Minimal image
+FROM base AS minimal
+ADD ./Gemfile /Gemfile
+ADD ./Gemfile.lock /Gemfile.lock
```
Comment (pepov) [inline v1.16/Dockerfile:51]: I don't see where is this used in this layer

Comment (OverOrion) [reply v1.16/Dockerfile]: Moved it to a stage later, thanks!

File: v1.16/Dockerfile (hunk 3 above)
Comment (OverOrion) [reply v1.16/Dockerfile]: Thanks, removed the duplicate package installs in [d96e27e](https://github.com/kube-logging/fluentd-images/pull/81/commits/d96e27e0430e03bbd826e047410236d3f6304f5e).

File: v1.16/Dockerfile (hunk 2 above)
Comment (OverOrion) [reply v1.16/Dockerfile]: [d96e27e](https://github.com/kube-logging/fluentd-images/pull/81/commits/d96e27e0430e03bbd826e047410236d3f6304f5e)

File: v1.16/Dockerfile (hunk 1 above)
Comment (OverOrion) [reply v1.16/Dockerfile]: [d96e27e](https://github.com/kube-logging/fluentd-images/pull/81/commits/d96e27e0430e03bbd826e047410236d3f6304f5e)

File: v1.16/Dockerfile (hunk 2 above)
Comment (OverOrion) [reply v1.16/Dockerfile]: I was a bit hasty, sorry!

The are installed and then removed at every stage, so each image is as small as possible, that's why it seemed like a duplicate, but it's not.

I am open to just installing it during the first stage and only removing it at the last stage, if you think it's not worth it.

Comment (pepov) [reply v1.16/Dockerfile]: depends on how big the result will be, but would make the dockerfile much simpler

Comment (OverOrion) [reply v1.16/Dockerfile]: The difference on the base target is 3 MBs:
```
$ docker images | grep clean

fluentd-base-without-cleaned-up               latest                               fa9756961272   4 seconds ago        189MB
fluentd-base-cleaned-up                       latest                               ca135c728c7a   About a minute ago   192MB

```
I think it makes sense to keep it in this case.

**Comment by OverOrion** [comment]
Superseded by https://github.com/kube-logging/fluentd-images/pull/82

**Pull request closed without merging**